### PR TITLE
Install logic ONLY supports our new installation format

### PR DESCRIPTION
**Pull request by MDrakos**
<table>
<td>
  <a href="https://activestatef.atlassian.net/browse/DX-790" title="DX-790" target="_blank">StoryDX-790</a>  Install logic ONLY supports our new installation format
  </td></table>
  <br />

### CODE REVIEW

File: cmd/state-installer/cmd.go [hunk 1]
```diff
@@ -250,6 +250,16 @@ func execute(out output.Outputer, cfg *config.Instance, an analytics.Dispatcher,
 		return errs.Wrap(err, "Could not detect if State Tool is already installed.")
 	}
 
+	if !stateToolInstalled && fileutils.DirExists(params.path) {
```
Comment (Naatan) [inline cmd/state-installer/cmd.go:253]: Please add a comment above, the installer in particular is very much geared towards outsiders reading this code.

File: cmd/state-installer/installer_lin_mac.go [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+	"io/ioutil"
+	"os"
+	"path/filepath"
+	"strings"
+
+	"github.com/ActiveState/cli/internal/errs"
+	"github.com/ActiveState/cli/internal/fileutils"
+)
+
+func (i *Installer) cleanInstallPath() error {
```
Comment (Naatan) [inline cmd/state-installer/installer_lin_mac.go:16]: We should make sure here that we don't need to worry about the additional application files that are created on macOS.

File: cmd/state-installer/installer_windows.go [hunk 3]
```diff
@@ -84,3 +72,67 @@ func (i *Installer) PrepareBinTargets(useBinDir bool) error {
 
 	return nil
 }
+
+func (i *Installer) cleanInstallPath() error {
+	if fileutils.DirExists(i.path) {
```
Comment (Naatan) [inline cmd/state-installer/installer_windows.go:77]: Return early

File: cmd/state-installer/installer_windows.go [hunk 4]
```diff
@@ -84,3 +72,67 @@ func (i *Installer) PrepareBinTargets(useBinDir bool) error {
...
+	if err != nil {
+		return errs.Wrap(err, "Could not determine if running as Windows administrator")
+	}
+
+	// Since we are repairing a corrupted install we need to also remove the old
+	// PATH entry. The new PATH entry will be added later in the install/update process.
+	// This is only an issue on Windows as on other platforms we can simply rewrite
+	// the PATH entry.
+	s := subshell.New(i.cfg)
+	if err := s.CleanUserEnv(i.cfg, sscommon.InstallID, !isAdmin); err != nil {
```
Comment (Naatan) [inline cmd/state-installer/installer_windows.go:107]: This should use the `isAdmin` value that we record at install time (ie. if the user installed State Tool as admin then we should also try to remove env vars here as admin, and if they aren't running as admin then we error out).

We should probably do this check WAY at the start of the updater. Meaning if you installed as admin then you HAVE to update as admin.

File: cmd/state-installer/installer_windows.go [hunk 5]
```diff
@@ -84,3 +72,67 @@ func (i *Installer) PrepareBinTargets(useBinDir bool) error {
...
+		return errs.Wrap(err, "Failed to State Tool installation PATH")
+	}
+
+	return nil
+}
+
+func removeOldExecutables(dir string) error {
+	if fileutils.TargetExists(dir) {
```
Comment (Naatan) [inline cmd/state-installer/installer_windows.go:115]: Return early

File: cmd/state-installer/installer_windows.go [hunk 6]
```diff
@@ -84,3 +72,67 @@ func (i *Installer) PrepareBinTargets(useBinDir bool) error {
...
+				continue
+			}
+
+			if strings.HasSuffix(file.Name(), ".old") {
+				logging.Debug("Deleting old file: %s", file.Name())
+				oldFile := filepath.Join(dir, file.Name())
+				if err := os.Remove(oldFile); err != nil {
+					multilog.Error("Failed to remove old executable: %s. Error: %s", oldFile, errs.JoinMessage(err))
+				}
+				continue
```
Comment (Naatan) [inline cmd/state-installer/installer_windows.go:132]: Doesn't serve any purpose.

File: internal/appinfo/appinfo.go [hunk 7]
```diff
@@ -22,6 +22,10 @@ type AppInfo struct {
 func execDir(baseDir ...string) (resultPath string) {
 	defer func() {
 		// Account for legacy use-case that wasn't using the correct bin dir
```
Comment (Naatan) [inline internal/appinfo/appinfo.go:24]: One of the purposes of this story is that we can get rid of this legacy handling logic. So I would expect to see this defer gone as part of this PR, rather than added to.

File: internal/installation/paths.go [hunk 8]
```diff
@@ -9,16 +10,23 @@ import (
...
 
-// CfgTransitionalStateToolPath is the configuration key for the path where a transitional State Tool might still be stored
-const CfgTransitionalStateToolPath = "transitional_installation_path"
+	// CfgTransitionalStateToolPath is the configuration key for the path where a transitional State Tool might still be stored
+	CfgTransitionalStateToolPath = "transitional_installation_path"
 
-const BinDirName = "bin"
+	BinDirName = "bin"
+
+	InstallDirMarker = ".install_root"
```
Comment (Naatan) [inline internal/installation/paths.go:26]: We should probably name this something more activestate specific. eg. `.state_install_root`.

File: internal/installation/paths.go [hunk 9]
```diff
@@ -63,11 +67,18 @@ func BinPathFromInstallPath(installPath string) (string, error) {
 }
 
 func InstalledOnPath(installPath string) (bool, string, error) {
```
Comment (Naatan) [inline internal/installation/paths.go:69]: This seems to be containing legacy logic again. As per our new logic the path would ALWAYS be `installPath/bin`, so we don't need to be looking at the StateApp exec path.

If we rely on this function for detecting legacy executables then we should split that off into separate logic that is ONLY used for the update logic. We don't want to sort of "infect" our main application logic with this legacy handling.

File: internal/installation/paths.go [hunk 10]
```diff
@@ -77,3 +88,65 @@ func LauncherInstallPath() (string, error) {
...
+	}
+
+	// Detect if the install dir has files in it
+	files, err := ioutil.ReadDir(path)
+	if err != nil {
+		return errs.Wrap(err, "Could not read directory: %s", path)
+	}
+
+	for _, file := range files {
+		if !file.IsDir() || strings.ToLower(file.Name()) != InstallDirMarker {
```
Comment (Naatan) [inline internal/installation/paths.go:128]: This is probably too aggressive. I think we can get by just checking if it has state executables in the wrong dir.

File: internal/locale/locales/en-us.yaml [hunk 11]
```diff
@@ -1835,6 +1835,13 @@ unstable_command_warning:
     This feature [ACTIONABLE]state {{.V0}}[/RESET] is still in beta. If you want to opt-in to unstable features, run the following command:
 
     [ACTIONABLE]state config set optin.unstable true[/RESET]
+err_update_corrupt_install:
```
Comment (Naatan) [inline internal/locale/locales/en-us.yaml:1838]: I don't think we actually use this, since the PR moved from erroring out to actually addressing the error. Though maybe we should print this as a fallback when the automatic cleanup fails.

File: test/integration/update_int_test.go [hunk 12]
```diff
@@ -143,6 +144,39 @@ func (suite *UpdateIntegrationTestSuite) testUpdate(ts *e2e.Session, baseDir str
...
+	ts.UseDistinctStateExes()
+
+	cfg, err := config.NewCustom(ts.Dirs.Config, singlethread.New(), true)
+	suite.Require().NoError(err)
+	defer cfg.Close()
+
+	subBinDir := filepath.Join(ts.Dirs.Bin, "bin")
+	ts.CopyExeToDir(ts.Exe, filepath.Join(subBinDir, filepath.Base(ts.Exe)))
+	ts.CopyExeToDir(ts.SvcExe, filepath.Join(subBinDir, filepath.Base(ts.SvcExe)))
+	ts.CopyExeToDir(ts.TrayExe, filepath.Join(subBinDir, filepath.Base(ts.TrayExe)))
```
Comment (Naatan) [inline test/integration/update_int_test.go:161]: Let's copy recursively from the dir that `ts.Exe` is in, that might help surface future edge-cases.

File: test/integration/install_scripts_int_test.go [hunk 13]
```diff
@@ -126,6 +127,33 @@ func (suite *InstallScriptsIntegrationTestSuite) TestInstall() {
...
+		)
+	} else {
+		cp = ts.SpawnCmdWithOpts("powershell.exe", e2e.WithArgs(argsWithActive...),
+			e2e.AppendEnv("SHELL="),
+			e2e.AppendEnv("ACTIVESTATE_CLI_DISABLE_RUNTIME=false"),
+		)
+	}
+	cp.ExpectLongString("Installation path must be an empty directory")
+	if runtime.GOOS != "windows" {
+		cp.ExpectExitCode(1)
```
Comment (Naatan) [inline test/integration/install_scripts_int_test.go:153]: Should file a bug; Windows should still exit non-zero.

**Review by Naatan — CHANGES_REQUESTED** [review]

File: cmd/state-installer/installer_lin_mac.go (hunk 2 above)
Comment (MDrakos) [reply cmd/state-installer/installer_lin_mac.go]: The system (launcher) files are reinstalled with every update. So we rewrite the install files including any symlinks so this shouldn't be an issue on macOS.

File: test/integration/update_int_test.go (hunk 12 above)
Comment (MDrakos) [reply test/integration/update_int_test.go]: Could not use `fileutils.CopyFiles` as it lead to recursion issues. Instead, I'm just reading the files in the `bin` directory and copying those.

File: test/integration/install_scripts_int_test.go (hunk 13 above)
Comment (MDrakos) [reply test/integration/install_scripts_int_test.go:153]: Created here: https://activestatef.atlassian.net/browse/DX-826

File: cmd/state-installer/installer.go [hunk 14]
```diff
@@ -160,6 +162,66 @@ func (i *Installer) sanitizeInput() error {
...
+		return errs.Wrap(errCorruptedInstall, "Bin path does not exist: %s", binPath)
+	}
+
+	// Detect if the install dir has files in it
+	files, err := ioutil.ReadDir(path)
+	if err != nil {
+		return errs.Wrap(err, "Could not read directory: %s", path)
+	}
+
+	for _, file := range files {
```
Comment (Naatan) [inline cmd/state-installer/installer.go:198]:
```suggestion
	// Executable files should be in bin dir, not root dir
	for _, file := range files {
```

File: cmd/state-installer/installer.go [hunk 15]
```diff
@@ -160,6 +162,66 @@ func (i *Installer) sanitizeInput() error {
...
+		return nil
+	}
+
+	// Detect if bin dir exists
+	binPath, err := installation.BinPathFromInstallPath(path)
+	if err != nil {
+		return errs.Wrap(err, "Could not detect bin path")
+	}
+	if !fileutils.DirExists(binPath) {
+		return errs.Wrap(errCorruptedInstall, "Bin path does not exist: %s", binPath)
```
Comment (Naatan) [inline cmd/state-installer/installer.go:189]: Going by the rest of the logic I don't think this actually matters.

File: internal/installation/paths.go [hunk 16]
```diff
@@ -66,20 +60,20 @@ func BinPathFromInstallPath(installPath string) (string, error) {
...
-	}
-
-	binPath, err := BinPathFromInstallPath(installPath)
-	if err != nil {
-		return false, "", errs.Wrap(err, "Could not detect binPath from BinPathFromInstallPath")
+func InstalledOnPath(installRoot string) (bool, string, error) {
+	var binPath string
+	var err error
+	if filepath.Base(installRoot) == "bin" {
+		binPath = installRoot
```
Comment (Naatan) [inline internal/installation/paths.go:67]: I think this is where we can use the InstallDirMarker. Rather than make assumptions about file paths we can just scan up the tree until we find InstallDirMarker.

File: internal/updater/updater.go [hunk 17]
```diff
@@ -126,6 +136,8 @@ func (u *AvailableUpdate) InstallBlocking(installTargetPath string, args ...stri
 	if u.Tag != nil {
 		envs = append(envs, fmt.Sprintf("%s=%s", constants.UpdateTagEnvVarName, *u.Tag))
 	}
+
+	logging.Debug("Executing: %s %s env: %s\n", installerPath, args, envs)
```
Comment (Naatan) [inline internal/updater/updater.go:140]: We're already debug logging this: https://github.com/ActiveState/cli/blob/3bf100013b38a2752b32c98192dde0dc40953860/internal/exeutils/exeutils.go#L120

**Review by Naatan — CHANGES_REQUESTED** [review]

File: internal/installation/paths.go [hunk 18]
```diff
@@ -62,13 +61,31 @@ func BinPathFromInstallPath(installPath string) (string, error) {
...
-func InstalledOnPath(installPath string) (bool, string, error) {
-	binPath, err := BinPathFromInstallPath(installPath)
+func InstalledOnPath(installRoot string) (bool, string, error) {
+	if !fileutils.DirExists(installRoot) {
+		return false, "", nil
+	}
+
+	path := appinfo.StateApp(installRoot).Exec()
+	if fileutils.TargetExists(path) {
+		return true, filepath.Dir(path), nil
```
Comment (MDrakos) [inline internal/installation/paths.go:71]: Right now I'm using the `InstallDirMarker` file as a fallback because older versions of the state tool will not have that file when updating.

File: cmd/state-installer/installer.go [hunk 19]
```diff
@@ -185,16 +185,14 @@ func detectCorruptedInstallDir(path string) error {
 	if err != nil {
 		return errs.Wrap(err, "Could not detect bin path")
 	}
-	if !fileutils.DirExists(binPath) {
```
Comment (Naatan) [inline cmd/state-installer/installer.go]: You'll have to account for this on line 203 where you try to read the `binPath`.

File: internal/installation/paths.go (hunk 18 above)
Comment (Naatan) [reply internal/installation/paths.go:71]: Please move any sort of logic dealing with old installation formats to the updater. The point of this story is that we can get rid of all this type of fallback / legacy logic and assume a pristine installation structure.

This could very well mean moving the entire `InstalledOnPath()` function into the updater/installer because that's the only place it's used. Point here is that the installer is the ONLY part of the code vulnerable to legacy or corrupted install dir formats.

**Review by Naatan — CHANGES_REQUESTED** [review]

File: cmd/state-installer/installer.go (hunk 19 above)
Comment (MDrakos) [reply cmd/state-installer/installer.go]: Don't we want to fail if the `bin` path doesn't exist? With the check removed this will fail if the `binPath` does not exist

Comment (MDrakos) [reply cmd/state-installer/installer.go]: ReadDir doesn't seems to return any identifiable error so I've added this check to the error case

File: cmd/state-installer/installer.go [hunk 20]
```diff
@@ -226,3 +229,30 @@ func isStateExecutable(name string) bool {
...
+	if !fileutils.DirExists(installRoot) {
+		return false, "", nil
+	}
+
+	path := appinfo.StateApp(installRoot).Exec()
+	if fileutils.TargetExists(path) {
+		return true, filepath.Dir(path), nil
+	}
+
+	found, err := fileutils.FindFileInPath(installRoot, installation.InstallDirMarker)
```
Comment (Naatan) [inline cmd/state-installer/installer.go:243]: I don't think this is where we should be using the InstallDirMarker. I would see it used in places like `appinfo.newAppInfo()` and `installation.InstallPath()`. This function is just checking whether the resulting install path has state tool on it.

**Review by Naatan — CHANGES_REQUESTED** [review]

File: cmd/state-installer/installer.go [hunk 21]
```diff
@@ -140,3 +161,98 @@ func (i *Installer) sanitize() error {
...
+	}
+
+	// Executable files should be in bin dir, not root dir
+	for _, file := range files {
+		if isStateExecutable(strings.ToLower(file.Name())) {
+			return errs.Wrap(errCorruptedInstall, "Install directory should only contain dirs: %s", path)
+		}
+	}
+
+	// Ensure that bin dir has at least the state and state-svc executables
```
Comment (Naatan) [inline cmd/state-installer/installer.go:202]: As discussed I think we can drop this as it's of no consequence.

**Review by Naatan — CHANGES_REQUESTED** [review]

File: cmd/state-installer/installer.go (hunk 20 above)
Comment (MDrakos) [reply cmd/state-installer/installer.go]: That's fair. I added it because of this comment: https://github.com/ActiveState/cli/pull/1784#discussion_r842183809 but the logic was different at the time.

**Review by Naatan — APPROVED** [review]
